### PR TITLE
Series decomposition transform with FFT

### LINKED ISSUE #425: Series decomposition transform with FFT
### 🚀 Task description

We might be interested in estimating the series decomposition to sinusoidal series. One of the scenarios, where this decomposition might be helpfull, is anomaly detection. This task proposes a new transform `FourierDecomposeTransform`, that will provide the ability of series decomposition in the pipeline.

### Plan

Proposed transform should work only with in-sample data.

Implement the following interface:

```python
class FourierDecomposeTransform(IrreversibleTransform):
    def __init__(self, k: int, in_column: str = "target", residuals: bool = False):
        pass

    def get_regressors_info(self) -> List[str]:
        return []

    def _fit(self, df: pd.DataFrame):
        pass

    def _transform(self, df: pd.DataFrame) -> pd.DataFrame:
        pass

    def fit(self, ts: TSDataset) -> "FourierDecomposeTransform":
        pass

    def transform(self, ts: TSDataset) -> TSDataset:
        pass
```

1. `__init__` - check initial conditions and initialize the object.
    * `k` - how many components to return. Must be greater than 0. If `int` provided must be lower than half of series length and return the specified number of frequncies.
    * `in_column` - which column will be used for decomposition
    * `residuals` - whether to add a column with residuals (e.g. target - sum(frequencies))
3. `get_regressors_info` - return an empty list, since transfrom doesn't produce any
4. `fit` - save timestamp bounds for the training data.
5. `transform`:
    * run timestamp checks to determine if in-sample data is being transformed; use padding with `nan` for timestamps in future
    * remove trailing and leading `nan`s from segment series
    * raise an error if missing values are presented
    * estimate Fourier representation of the series
    * select positive frequencies (first half)
    * select top-k frequencies proportional to amplitudes
    * return frequencies to the real domain with the inverse Fourier transform
    * pad with `nan`s to match the original series length if needed 
    * estimate residuals if necessary.
    * add decomposition to the dataset

### Test cases

1. Test in-sample and out-of-sample cases
2. Add tests similar to other transforms (inference test, etc.).
3. Test that the transform works with different cases for `k` (`int`, edge cases)
4. Test that transform works in the pipeline with other transforms (especially with `IForestOutlierTransform`)
5. Ensure decomposition is not corrupted or lost after calling (`pipelin.forecast`, `pipeline.predict` with components and intervals)
6. Ensure backtest works as expected

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by brsnw250**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/etna-team/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/etna-team/etna/blob/master/CHANGELOG.md)?

## Closing issues

closes #425

### CODE REVIEW

File: tests/test_transforms/test_decomposition/test_dft_based.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+def test_timestamp_from_history(ts_with_exogs_train_test):
+    test, train = ts_with_exogs_train_test
+    transform = FourierDecomposeTransform(k=5)
+    transform.fit_transform(train)
+
+    with pytest.raises(ValueError, match="First index of the dataset to be transformed must be larger"):
+        transform.transform(test)
+
+
+def test_timestamp_from_future(ts_with_exogs_train_test):
```
Comment (d-a-bunin) [inline tests/test_transforms/test_decomposition/test_dft_based.py:181]: I can't see any difference with a previous test, why the results are different?

File: tests/test_transforms/test_decomposition/test_dft_based.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+    (
+        ("outliers_df_with_two_columns", "target"),
+        ("outliers_df_with_two_columns", "feature"),
+        ("ts_with_exogs", "target"),
+        ("ts_with_exogs", "exog"),
+    ),
+)
+@pytest.mark.parametrize("k", (1, 5, 10, 40, 51))
+@pytest.mark.parametrize("forecast_model", (ProphetModel(), CatBoostPerSegmentModel(iterations=10)))
+def test_pipeline_ks(ts_name, in_column, forecast_model, k, request):
```
Comment (d-a-bunin) [inline tests/test_transforms/test_decomposition/test_dft_based.py:284]: What `ks` mean in the name of the test?

File: tests/test_transforms/test_decomposition/test_dft_based.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+            TimeSeriesImputerTransform(in_column="target"),
+        ],
+        model=ProphetModel(),
+        horizon=3,
+    )
+    pipeline.fit(ts)
+
+
+@pytest.mark.parametrize("k", (1, 5))
+def test_stride_transform(forward_stride_datasets, k):
```
Comment (d-a-bunin) [inline tests/test_transforms/test_decomposition/test_dft_based.py:335]: I don't really understand what we are testing here.

File: tests/test_transforms/test_inference/test_transform.py [hunk 4]
```diff
@@ -2420,6 +2434,7 @@ def _test_transform_future_with_target(self, ts, transform, expected_changes, ga
                 "regular_ts",
                 {"create": {"res"}},
             ),
+            # (FourierDecomposeTransform(in_column="target", k=2, residuals=True), "regular_ts",             {"create": {"target_dft_0", "target_dft_1", "target_dft_residuals"}}),
```
Comment (d-a-bunin) [inline tests/test_transforms/test_inference/test_transform.py:2437]: Why is this commented?

File: tests/test_transforms/test_inference/test_inverse_transform.py [hunk 5]
```diff
@@ -2589,6 +2595,7 @@ def _test_inverse_transform_future_with_target(
                 "regular_ts",
                 {},
             ),
+            # (FourierDecomposeTransform(in_column="target", k=5, residuals=True), "regular_ts", {}),
```
Comment (d-a-bunin) [inline tests/test_transforms/test_inference/test_inverse_transform.py:2598]: Why is this commented?

File: etna/transforms/decomposition/dft_based.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        components = np.fft.ifft(selected_freqs).real
+
+        components_df = pd.DataFrame(
+            data=components.T, columns=[f"dft_{i}" for i in range(components.shape[0])], index=series.index
+        )
+
+        if self.residuals:
+            components_df["dft_residuals"] = series.values - np.sum(components, axis=0)
+
+        if not components_df.index.equals(initial_index):
```
[the remark this answers is not in the document]
Comment (d-a-bunin) [reply etna/transforms/decomposition/dft_based.py:120]: Can we add some comments about this cases?

Why do we need this if at all? Can't we just use the expression as is every time?

File: tests/test_transforms/test_decomposition/test_dft_based.py [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+
+
+def test_check_segments_missing_values(ts_with_missing):
+    df = ts_with_missing[..., "target"]
+    transform = FourierDecomposeTransform(k=5)
+    with pytest.raises(ValueError, match=f"Feature `target` contains missing values"):
+        transform._check_segments(df=df)
+
+
+def test_check_segments_large_k(ts_with_exogs):
```
[the remark this answers is not in the document]
Comment (d-a-bunin) [reply tests/test_transforms/test_decomposition/test_dft_based.py:100]: Now we are setting `k=100`, could we also (using parametrization) set some minimal value that leads to error on `ts_with_exogs`? It could better detect test regression in the future.

File: tests/test_transforms/test_inference/test_transform.py (hunk 4 above)
Comment (d-a-bunin) [reply tests/test_transforms/test_inference/test_transform.py:2437]: It is still relevant.

File: tests/test_transforms/test_decomposition/test_dft_based.py (hunk 1 above)
Comment (d-a-bunin) [reply tests/test_transforms/test_decomposition/test_dft_based.py:179]: How are they different if we are using the same fixture?

File: etna/transforms/decomposition/dft_based.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+        # compute Fourier decomposition of the series
+        dft_series = np.fft.fft(series)
+
+        # compute "amplitudes" for each frequency
+        abs_dft_series = np.abs(dft_series)
+
+        # select top-k indices
+        abs_pos_dft_series = abs_dft_series[:num_pos_freqs]
+        top_k_idxs = np.argpartition(abs_pos_dft_series, num_pos_freqs - self.k)[-self.k :]
```
[the remark this answers is not in the document]
Comment (d-a-bunin) [reply etna/transforms/decomposition/dft_based.py:103]: Can this lead to different results every run of the `transform` even on the same data?

Current behaviour seems problematic.

[the remark this answers is not in the document]
Comment (brsnw250) [reply etna/transforms/decomposition/dft_based.py:103]: No, same data will result in the same components

File: tests/test_transforms/test_decomposition/test_dft_based.py (hunk 1 above)
Comment (brsnw250) [reply tests/test_transforms/test_decomposition/test_dft_based.py:179]: Different order
`test, train = ts_with_exogs_train_test` in the previos test
`train, test = ts_with_exogs_train_test` here

File: tests/test_transforms/test_decomposition/test_dft_based.py [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+def test_stride_transform(forward_stride_datasets, k):
+    train, test = forward_stride_datasets
+
+    transform = FourierDecomposeTransform(k=k, residuals=True)
+
+    transform.fit(train)
+    transformed = transform.transform(test)
+
+    assert not transformed.df.iloc[:10].isna().any().any()
+    assert transformed.df.iloc[10:].isna().all().any()
```
[the remark this answers is not in the document]
Comment (brsnw250) [reply tests/test_transforms/test_decomposition/test_dft_based.py:342]: No, this checks `nan`s are added by transform, other columns are ok

File: tests/test_transforms/test_inference/test_inverse_transform.py (hunk 5 above)
Comment (brsnw250) [reply tests/test_transforms/test_inference/test_inverse_transform.py:2598]: In such cases dataset for transform has only timestamps from future. So error will be raised and test would never pass.

File: tests/test_transforms/test_decomposition/test_dft_based.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (brsnw250) [reply tests/test_transforms/test_decomposition/test_dft_based.py]: Added test

**Review by d-a-bunin — APPROVED** [review]

**Review by d-a-bunin — APPROVED** [review]
